File: gherkin/dart/lib/src/ast/Examples.dart
import 'package:gherkin/core.dart';
import 'package:gherkin/extensions.dart';
import 'package:gherkin/language.dart';

import 'IHasDescription.dart';
import 'IHasLocation.dart';
import 'IHasRows.dart';
import 'IHasTags.dart';
import 'TableRow.dart';
import 'Tag.dart';

class Examples implements IHasLocation, IHasDescription, IHasRows, IHasTags
    , INullSafetyObject
{
  static Examples empty = _InvalidExamples();

  final Iterable<Tag> _tags;
  final Location _location;
  final String _keyword;
  final String _name;
  final String _description;
  final TableRow _tableHeader;
  final Iterable<TableRow> _tableBody;
  late final List<TableRow> _rows;

  @override
  Iterable<Tag> get tags => _tags;
  @override
  Location get location => _location;
  @override
  String get keyword => _keyword;
  @override
  String get name => _name;
  @override
  String get description => _description;

  TableRow get tableHeader => _tableHeader;

  Iterable<TableRow> get tableBody => _tableBody;

  Examples(this._tags, this._location, this._keyword, this._name
      , this._description, this._tableHeader, this._tableBody)
  {
    _rows = <TableRow>[];
    if( _tableHeader.isNotEmpty ) {
      _rows.add(_tableHeader);
    }
    if( _tableBody.isNotEmpty ) {
      _rows.addAll(tableBody);
    }
  }

  @override
  Iterable<TableRow> get rows => _rows;

  @override
  bool get isEmpty => false;/*_tags.isEmpty
      && _location.isEmpty
      && _keyword.isEmpty
      && _name.isEmpty
      && _description.isEmpty
      && _tableHeader.isEmpty
      && _tableBody.isEmpty
      && _rows.isEmpty;*/

  @override
  bool get isNotEmpty => !isEmpty;
}


/// Convenience implementation of an invalid [Examples] instance.
class _InvalidExamples extends Examples
{
  _InvalidExamples() : super(const <Tag>[], Location.empty, Strings.empty
    , Strings.empty, Strings.empty, TableRow.empty, const <TableRow>[] );

  @override
  bool get isEmpty => true;
}

File: gherkin/dotnet/Gherkin/Ast/Examples.cs
using System.Collections.Generic;
using System.Linq;

namespace Gherkin.Ast
{
    public class Examples : IHasLocation, IHasDescription, IHasRows, IHasTags
    {
        public IEnumerable<Tag> Tags { get; private set; }
        public Location Location { get; private set; }
        public string Keyword { get; private set; }
        public string Name { get; private set; }
        public string Description { get; private set; }
        public TableRow TableHeader { get; private set; }
        public IEnumerable<TableRow> TableBody { get; private set; }

        public Examples(Tag[] tags, Location location, string keyword, string name, string description, TableRow header, TableRow[] body)
        {
            Tags = tags;
            Location = location;
            Keyword = keyword;
            Name = name;
            Description = description;
            TableHeader = header;
            TableBody = body;
        }

        IEnumerable<TableRow> IHasRows.Rows
        {
            get { return new TableRow[] {TableHeader}.Concat(TableBody); }
        }
    }
}